src/Roslyn.Diagnostics.Analyzers/UnitTests/DeclarePublicAPIAnalyzerTests.cs

```diff
@@ -518,5 +538,365 @@ public class C
+    {
+        public int Field;
+    }
+}
+
+public class C2 { }
+";
+
+            var shippedText = @"C.CC
+C.CC.CC() -> void";
+            var unshippedText = @"";
+            var fixedUnshippedText = @"C
+C.CC.Field -> int
+C.Field -> int
+C2
+C2.C2() -> void";
+
+            VerifyCSharpAdditionalFileFix(source, shippedText, unshippedText, fixedUnshippedText);
+        }
+
+        [Fact]
+        public void TestMissingNestedGenericMembersAndStaleMembers_Fix()
+        {
+            var source = @"
+public class C
+{
+    private C() { }
+    public CC<int> Field;
+    private C3.C4 Field2;
+    private C3.C4 Method(C3.C4 p1) { }
+
+    public class CC<T>
+    {
+        public int Field;
+        public CC<int> Field2;
+    }
+    
+    public class C3
+    {
+        public class C4 { }
+    }
+}
+
+public class C2 { }
+";
+
+            var shippedText = @"";
+            var unshippedText = @"C.C3
+C.C3.C3() -> void
+C.C3.C4
+C.C3.C4.C4() -> void
+C.CC<T>
+C.CC<T>.CC() -> void
+C.Field2 -> C.C3.C4
+C.Method(C.C3.C4 p1) -> C.C3.C4
+";
+            var fixedUnshippedText = @"C
+C.C3
+C.C3.C3() -> void
+C.C3.C4
+C.C3.C4.C4() -> void
+C.CC<T>
+C.CC<T>.CC() -> void
+C.CC<T>.Field -> int
+C.CC<T>.Field2 -> C.CC<int>
+C.Field -> C.CC<int>
+C2
+C2.C2() -> void";
+
+            VerifyCSharpAdditionalFileFix(source, shippedText, unshippedText, fixedUnshippedText);
+        }
+
+        [Fact]
+        public void TestWithExistingUnshippedNestedMembers_Fix()
+        {
+            var source = @"
+public class C
+{
+    private C() { }
+    public int Field;
+
+    public class CC
+    {
+        public int Field;
+    }
+}
+
+public class C2 { }
+";
+
+            var shippedText = @"";
+            var unshippedText = @"C.CC
+C.CC.CC() -> void
+C.CC.Field -> int";
+            var fixedUnshippedText = @"C
+C.CC
+C.CC.CC() -> void
+C.CC.Field -> int
+C.Field -> int
+C2
+C2.C2() -> void";
+
+            VerifyCSharpAdditionalFileFix(source, shippedText, unshippedText, fixedUnshippedText);
+        }
+
+        [Fact]
+        public void TestWithExistingShippedNestedMembers_Fix()
+        {
+            var source = @"
+public class C
+{
+    private C() { }
+    public int Field;
+
+    public class CC
+    {
+        public int Field;
+    }
+}
+
+public class C2 { }
+";
+
+            var shippedText = @"C.CC
+C.CC.CC() -> void
+C.CC.Field -> int";
+            var unshippedText = @"";
+            var fixedUnshippedText = @"C
+C.Field -> int
+C2
+C2.C2() -> void";
+
+            VerifyCSharpAdditionalFileFix(source, shippedText, unshippedText, fixedUnshippedText);
+        }
+
+        [Fact]
+        public void TestOnlyRemoveStaleSiblingEntries_Fix()
+        {
+            var source = @"
+public class C
+{
+    private C() { }
+    public int Field;
+
+    public class CC
+    {
+        private int Field; // This has a stale public API entry, but this shouldn't be removed unless we attempt to add a public API entry for a sibling.
```

Comment:
So the term stale entry is probably a misnomer. We want to remove the stale entries for siblings when doing a signature change for unshipped public API. However, we still want the user to be explicit about removing stale entries from other unrelated types in the compilation - user might have accidentally deleted/changed another unshipped public API, and it would be weird if its public API entry gets cleared when trying to add an entry for a new symbol in completely unrelated type.